### ISSUE TITLE
Missing combine of shl+bswap to rol

### ISSUE THREAD

**Issue by davidbolvansky** [llvm/llvm-project#51391]
|  |  |
| --- | --- |
| Bugzilla Link | [52049](https://llvm.org/bz52049) |
| Version | trunk |
| OS | Linux |
| CC | @RKSimon,@rotateright |

## Extended Description
```
unsigned read_be16 (unsigned char *data)
{
  return *(data + 1) | (*data << 8);
}

LLVM
read_be16(unsigned char*):
        movzx   eax, word ptr [rdi]
        shl     eax, 16
        bswap   eax
        ret

GCC
read_be16(unsigned char*):
        movzx   eax, WORD PTR [rdi]
        rol     ax, 8
        movzx   eax, ax
        ret
```

https://godbolt.org/z/s8ha55Y7W

**Comment by davidbolvansky** [comment]
Yeah, this is backend specific issue. Still happens with trunk.

**Comment by RKSimon** [comment]
Candidate Patch: https://reviews.llvm.org/D120192

**Comment by RKSimon** [comment]
Fixed in 370ebc9d9a573d6a0bcb0d7c3e0a57d55b7a7673
